cpp/src/join/hash_join.cu

```diff
@@ -87,22 +90,37 @@ std::size_t compute_join_output_size(table_device_view build_table,
   }
 
   auto const probe_nulls = cudf::nullate::DYNAMIC{has_nulls};
-  pair_equality equality{probe_table, build_table, probe_nulls, nulls_equal};
 
-  row_hash hash_probe{probe_nulls, probe_table};
+  auto row_hash                 = cudf::experimental::row::hash::row_hasher{preprocessed_probe};
+  auto hash_probe               = row_hash.device_hasher(probe_nulls);
   auto const empty_key_sentinel = hash_table.get_empty_key_sentinel();
   make_pair_function pair_func{hash_probe, empty_key_sentinel};
 
-  auto iter = cudf::detail::make_counting_transform_iterator(0, pair_func);
+  auto const row_comparator =
+    cudf::experimental::row::equality::two_table_comparator{preprocessed_probe, preprocessed_build};
 
-  std::size_t size;
-  if constexpr (JoinKind == join_kind::LEFT_JOIN) {
-    size = hash_table.pair_count_outer(iter, iter + probe_table_num_rows, equality, stream.value());
+  auto const comparator_helper = [&](auto const device_comparator) {
+    pair_equality equality{device_comparator};
+
+    std::size_t size;
+    auto iter = cudf::detail::make_counting_transform_iterator(0, pair_func);
+    if constexpr (JoinKind == join_kind::LEFT_JOIN) {
```

Comment:
I also notice that the function template parameter `JoinKind` is used for very simple conditions. However, with that template the entire function will be instantiated multiple times, which is very expensive. That is the reason why compiling this file is very slow. I would recommend removing the template parameter entirely. That should have negligibly small impact on run time performance since runtime `if/else` conditions cost just nano-seconds.